picard/webservice/throttle.py

```diff
@@ -0,0 +1,205 @@
+# Plugins may assign limits to their associated service(s) like so:
+#
+# >>> from picard.webservice.throttle import set_minimum_request_delay
+# >>> set_minimum_request_delay(('myservice.org', 80), 100)  # 10 requests/second
+REQUEST_DELAY_MINIMUM = defaultdict(lambda: 1000)
+
+# Current delay (adaptive) between requests to a given hostkey.
+REQUEST_DELAY = defaultdict(lambda: 1000)  # Conservative initial value.
+
+# Determines delay during exponential backoff phase.
+REQUEST_DELAY_EXPONENT = defaultdict(lambda: 0)
+
+# Unacknowledged request counter.
+#
+# Bump this when handing a request to QNetworkManager and trim when receiving
+# a response.
+CONGESTION_UNACK = defaultdict(lambda: 0)
+
+# Congestion window size in terms of unacked requests.
+#
+# We're allowed to send up to `int(this)` many requests at a time.
+CONGESTION_WINDOW_SIZE = defaultdict(lambda: 1.0)
+
+# Slow start threshold.
+#
+# After placing this many unacknowledged requests on the wire, switch from
+# slow start to congestion avoidance.  (See `_adjust_throttle`.)  Initialized
+# upon encountering a temporary error.
+CONGESTION_SSTHRESH = defaultdict(lambda: 0)
+
+# Storage of last request times per host key
+LAST_REQUEST_TIMES = defaultdict(lambda: 0)
+
+
+def set_minimum_request_delay(hostkey, value):
+    REQUEST_DELAY_MINIMUM[hostkey] = value
+
+
+def current_delay(hostkey):
+    return REQUEST_DELAY[hostkey]
+
+
+def get_delay_to_next_request(hostkey):
+    """Calculate delay to next request to hostkey (host, port)
+       returns a tuple (wait, delay) where:
+           wait is True if a delay is needed
+           delay is the delay in milliseconds to next request
+    """
+
+    if CONGESTION_UNACK[hostkey] >= int(CONGESTION_WINDOW_SIZE[hostkey]):
+        # We've maxed out the number of requests to `hostkey`, so wait
+        # until responses begin to come back.  (See `_timer_run_next_task`
+        # strobe in `_handle_reply`.)
+        return (True, sys.maxsize)
+
+    interval = REQUEST_DELAY[hostkey]
+    if not interval:
+        log.debug("throttle: Starting another request to %s without delay", hostkey)
+        return (False, 0)
+    last_request = LAST_REQUEST_TIMES[hostkey]
+    if not last_request:
+        log.debug("throttle: First request to %s", hostkey)
+        _remember_request_time(hostkey)  # set it on first run
+        return (False, interval)
+    elapsed = (time.time() - last_request) * 1000
+    if elapsed >= interval:
+        log.debug(
+            "throttle: Last request to %s was %d ms ago, starting another one", hostkey, elapsed)
+        return (False, interval)
+    delay = int(math.ceil(interval - elapsed))
+    log.debug("throttle: Last request to %s was %d ms ago, waiting %d ms before starting another one",
+              hostkey, elapsed, delay)
+    return (True, delay)
+
+
+def _remember_request_time(hostkey):
+    if REQUEST_DELAY[hostkey]:
+        LAST_REQUEST_TIMES[hostkey] = time.time()
+
+
+def incr_requests(hostkey):
```

Comment:
Do not like the function name that much. Maybe `increment_unacked_requests`? Or is that too long?

Reply:
> Do not like the function name that much. Maybe increment_unacked_requests? Or is that too long?

I don't like it much either, and yes,  `increment_unacked_requests` is a bit long.
Why not `increment_requests` and `decrement_requests` ? mid-sized & explicit.